File: apps/dsa-web/src/api/portfolioHealth.ts
// Copyright (c) 2026 SiinXu / StockPulse contributors
// SPDX-License-Identifier: AGPL-3.0-only
// Typed client for portfolio health read and explicit refresh operations.

import { z } from 'zod';
import apiClient, { locallyRecoverableResourceConfig } from './index';
import { parseCamelCasePayload } from './parseCamelCasePayload';
import { createApiError, createParsedApiError } from './error';
import type { operations } from '../types/api.generated';
import type {
  PortfolioHealthQuery,
  PortfolioHealthRefreshQuery,
  PortfolioHealthResponse,
} from '../types/portfolioHealth';

type OpenApiGetQuery = NonNullable<operations['getPortfolioHealth']['parameters']['query']>;
type OpenApiRefreshQuery = NonNullable<operations['refreshPortfolioHealth']['parameters']['query']>;
const _getQueryAnchor: keyof OpenApiGetQuery = 'cost_method';
const _refreshQueryAnchor: keyof OpenApiRefreshQuery = 'persist';
void _getQueryAnchor;
void _refreshQueryAnchor;

const finiteNumber = z.number().refine(Number.isFinite, {
  message: 'non-finite number rejected',
});
const optionalFinite = finiteNumber.nullable().optional();

const portfolioHealthBandSchema = z.enum(['healthy', 'fair', 'caution', 'poor']);
const portfolioHealthStatusSchema = z.enum([
  'ok',
  'partial',
  'empty_portfolio',
  'unavailable',
]);

const portfolioHealthDimensionNameSchema = z.enum([
  'concentration',
  'risk_exposure',
  'diversification',
  'pnl',
  'cash_ratio',
]);

const weightsSchema = z.object({
  concentration: finiteNumber,
  riskExposure: finiteNumber,
  diversification: finiteNumber,
  pnl: finiteNumber,
  cashRatio: finiteNumber,
});

const effectiveWeightsSchema = z.object({
  concentration: optionalFinite,
  riskExposure: optionalFinite,
  diversification: optionalFinite,
  pnl: optionalFinite,
  cashRatio: optionalFinite,
});

const dimensionSchema = z.object({
  formula: z.string().nullable().optional(),
  input: z.record(z.string(), finiteNumber).optional(),
  reason: z.string().nullable().optional(),
  score: optionalFinite,
  status: z.enum(['ok', 'unavailable']),
  statusMessage: z.string().nullable().optional(),
}).passthrough();

export const portfolioHealthResponseSchema = z
  .object({
    accountId: z.number().int().nullable().optional(),
    asOf: z.string().min(1),
    band: portfolioHealthBandSchema.nullable().optional(),
    bands: z.array(z.object({
      maxExclusive: finiteNumber,
      minInclusive: finiteNumber,
      name: portfolioHealthBandSchema,
    }).passthrough()).optional().default([]),
    comparable: z.boolean(),
    config: z.object({
      cashHighAlertPct: finiteNumber,
      cashLowAlertPct: finiteNumber,
      concentrationAlertPct: finiteNumber,
      diversificationAlert: finiteNumber,
      pnlLossAlertPct: finiteNumber,
      source: z.literal('shared_config'),
      varAlertPct: finiteNumber,
      weights: weightsSchema,
    }).passthrough(),
    costMethod: z.enum(['fifo', 'avg']),
    coverageRatio: finiteNumber,
    currency: z.string().min(1),
    dataQuality: z.object({
      fxStale: z.boolean(),
      limitations: z.array(z.string()).optional().default([]),
      missingPriceSymbols: z.array(z.string()).optional().default([]),
      partialReasons: z.array(z.string()).optional().default([]),
      riskMetricsStatus: z.string().nullable().optional(),
      snapshotDataQuality: z.string().nullable().optional(),
      status: z.enum(['ok', 'partial', 'empty', 'unavailable']),
    }).passthrough(),
    dimensions: z.object({
      concentration: dimensionSchema,
      riskExposure: dimensionSchema,
      diversification: dimensionSchema,
      pnl: dimensionSchema,
      cashRatio: dimensionSchema,
    }),
    status: portfolioHealthStatusSchema,
    statusMessage: z.string().nullable().optional(),
    disclaimer: z.string(),
    effectiveWeights: effectiveWeightsSchema,
    formulaVersion: z.literal('portfolio_health_v2'),
    inputs: z.object({
      cashPct: optionalFinite,
      diversificationScore: optionalFinite,
      topWeightPct: optionalFinite,
      totalCash: finiteNumber,
      totalEquity: finiteNumber,
      totalMarketValue: finiteNumber,
      unrealizedPnlPct: optionalFinite,
      varPct: optionalFinite,
    }).passthrough(),
    insights: z.array(z.object({
      code: z.string(),
      message: z.string(),
      metric: z.string().nullable().optional(),
      severity: z.enum(['info', 'warning']),
      source: z.enum(['rule', 'rule+llm_polish']),
      symbol: z.string().nullable().optional(),
      threshold: optionalFinite,
      value: optionalFinite,
    }).passthrough()).optional().default([]),
    llmCanModifyScore: z.literal(false),
    partialScore: optionalFinite,
    persisted: z.boolean(),
    provenance: z.object({
      calculatedAt: z.string().min(1),
      configHash: z.string(),
      fxProvenance: z.record(z.string(), z.unknown()).optional(),
      priceProvenance: z.record(z.string(), z.unknown()).optional(),
      riskHash: z.string(),
      riskHistory: z.record(z.string(), z.unknown()).optional(),
      snapshotHash: z.string(),
    }).passthrough(),
    score: optionalFinite,
    scoreSource: z.literal('rules'),
    unavailableDimensions: z.array(portfolioHealthDimensionNameSchema).optional().default([]),
    weights: weightsSchema,
  })
  .passthrough();

export const portfolioHealthQuerySchema = z.object({
  accountId: z.number().int().positive().optional(),
  asOf: z.string().min(1).optional(),
  costMethod: z.enum(['fifo', 'avg']).optional(),
}).strict();

const portfolioHealthRefreshQuerySchema = portfolioHealthQuerySchema.extend({
  persist: z.boolean().optional(),
});

function invalidQueryError(issues: z.ZodIssue[]): never {
  const summary = issues
    .map((issue) => `${issue.path.join('.') || '(root)'}: ${issue.message}`)
    .join('; ');
  throw createApiError(createParsedApiError({
    title: '请求参数无效',
    message: `Portfolio health query failed validation. ${summary}`,
    rawMessage: summary,
    category: 'missing_params',
    code: 'invalid_params',
    params: { issues: summary },
    details: issues,
  }));
}

function buildParams(query: PortfolioHealthRefreshQuery): Record<string, string | number | boolean> {
  const params: Record<string, string | number | boolean> = {};
  if (query.accountId != null) params.account_id = query.accountId;
  if (query.asOf) params.as_of = query.asOf;
  if (query.costMethod) params.cost_method = query.costMethod;
  if (query.persist != null) params.persist = query.persist;
  return params;
}

function parseResponse(data: unknown): PortfolioHealthResponse {
  const parsed = parseCamelCasePayload<PortfolioHealthResponse>(
    data,
    portfolioHealthResponseSchema,
    'PortfolioHealthResponse',
    'portfolioHealth',
  );
  return {
    ...parsed,
    bands: Array.isArray(parsed.bands) ? parsed.bands : [],
    dataQuality: {
      ...parsed.dataQuality,
      limitations: Array.isArray(parsed.dataQuality.limitations)
        ? parsed.dataQuality.limitations
        : [],
      missingPriceSymbols: Array.isArray(parsed.dataQuality.missingPriceSymbols)
        ? parsed.dataQuality.missingPriceSymbols
        : [],
      partialReasons: Array.isArray(parsed.dataQuality.partialReasons)
        ? parsed.dataQuality.partialReasons
        : [],
    },
    insights: Array.isArray(parsed.insights) ? parsed.insights : [],
    unavailableDimensions: Array.isArray(parsed.unavailableDimensions)
      ? parsed.unavailableDimensions
      : [],
  };
}

export const portfolioHealthApi = {
  /**
   * Read-only stored daily snapshot. Never computes or persists.
   * 404 means no snapshot yet (empty / not refreshed).
   */
  async getSummary(query: PortfolioHealthQuery = {}): Promise<PortfolioHealthResponse | null> {
    const result = portfolioHealthQuerySchema.safeParse(query);
    if (!result.success) return invalidQueryError(result.error.issues);
    try {
      const response = await apiClient.get<unknown>(
        '/api/v1/portfolio/health',
        { ...locallyRecoverableResourceConfig(), params: buildParams(result.data) },
      );
      return parseResponse(response.data);
    } catch (error) {
      const status = (error as { response?: { status?: number } })?.response?.status;
      if (status === 404) return null;
      throw error;
    }
  },

  /** Compute and persist the current daily snapshot after an explicit user action. */
  async refresh(query: PortfolioHealthRefreshQuery = {}): Promise<PortfolioHealthResponse> {
    const result = portfolioHealthRefreshQuerySchema.safeParse(query);
    if (!result.success) return invalidQueryError(result.error.issues);
    const response = await apiClient.post<unknown>(
      '/api/v1/portfolio/health/refresh',
      undefined,
      { params: buildParams({ ...result.data, persist: result.data.persist ?? true }) },
    );
    return parseResponse(response.data);
  },
};

File: src/api/v1/schemas/portfolio_health.py
# -*- coding: utf-8 -*-
"""Strict schemas for deterministic portfolio health snapshots."""

from __future__ import annotations

from datetime import date, datetime
from typing import Annotated, Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


FiniteFloat = Annotated[float, Field(allow_inf_nan=False)]
ScoreFloat = Annotated[float, Field(ge=0.0, le=100.0, allow_inf_nan=False)]
UnitFloat = Annotated[float, Field(ge=0.0, le=1.0, allow_inf_nan=False)]
HealthBandName = Literal["healthy", "fair", "caution", "poor"]
HealthStatus = Literal["ok", "partial", "empty_portfolio", "unavailable"]
DimensionStatus = Literal["ok", "unavailable"]


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)


class PortfolioHealthBand(_StrictModel):
    name: HealthBandName
    min_inclusive: ScoreFloat
    max_exclusive: ScoreFloat


class PortfolioHealthInsight(_StrictModel):
    code: str = Field(min_length=1, max_length=96)
    severity: Literal["info", "warning"]
    message: str = Field(min_length=1, max_length=1000)
    symbol: Optional[str] = Field(default=None, max_length=32)
    metric: Optional[str] = Field(default=None, max_length=96)
    value: Optional[FiniteFloat] = None
    threshold: Optional[FiniteFloat] = None
    source: Literal["rule", "rule+llm_polish"] = "rule"


class PortfolioHealthDataQuality(_StrictModel):
    status: Literal["ok", "partial", "empty", "unavailable"]
    fx_stale: bool = False
    snapshot_data_quality: Optional[str] = Field(default=None, max_length=64)
    limitations: List[str] = Field(default_factory=list, max_length=256)
    missing_price_symbols: List[str] = Field(default_factory=list, max_length=256)
    risk_metrics_status: Optional[str] = Field(default=None, max_length=64)
    partial_reasons: List[str] = Field(default_factory=list, max_length=64)


class PortfolioHealthInputs(_StrictModel):
    top_weight_pct: Optional[ScoreFloat] = None
    var_pct: Optional[ScoreFloat] = None
    diversification_score: Optional[UnitFloat] = None
    unrealized_pnl_pct: Optional[FiniteFloat] = None
    cash_pct: Optional[FiniteFloat] = None
    total_equity: FiniteFloat = 0.0
    total_cash: FiniteFloat = 0.0
    total_market_value: FiniteFloat = 0.0


class PortfolioHealthDimension(_StrictModel):
    status: DimensionStatus
    score: Optional[ScoreFloat] = None
    input: Dict[str, FiniteFloat] = Field(default_factory=dict, max_length=8)
    formula: Optional[str] = Field(default=None, max_length=500)
    reason: Optional[str] = Field(default=None, max_length=96)
    status_message: Optional[str] = Field(default=None, max_length=500)


class PortfolioHealthDimensions(_StrictModel):
    concentration: PortfolioHealthDimension
    risk_exposure: PortfolioHealthDimension
    diversification: PortfolioHealthDimension
    pnl: PortfolioHealthDimension
    cash_ratio: PortfolioHealthDimension


class PortfolioHealthWeights(_StrictModel):
    concentration: UnitFloat
    risk_exposure: UnitFloat
    diversification: UnitFloat
    pnl: UnitFloat
    cash_ratio: UnitFloat


class PortfolioHealthEffectiveWeights(_StrictModel):
    concentration: Optional[UnitFloat] = None
    risk_exposure: Optional[UnitFloat] = None
    diversification: Optional[UnitFloat] = None
    pnl: Optional[UnitFloat] = None
    cash_ratio: Optional[UnitFloat] = None


class PortfolioHealthResolvedConfig(_StrictModel):
    weights: PortfolioHealthWeights
    concentration_alert_pct: ScoreFloat
    cash_low_alert_pct: ScoreFloat
    cash_high_alert_pct: ScoreFloat
    var_alert_pct: ScoreFloat
    diversification_alert: UnitFloat
    pnl_loss_alert_pct: Annotated[
        float, Field(ge=-100.0, le=0.0, allow_inf_nan=False)
    ]
    source: Literal["shared_config"]


class PortfolioHealthProvenance(_StrictModel):
    snapshot_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    risk_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    config_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    calculated_at: datetime
    risk_history: Dict[str, Any] = Field(default_factory=dict)
    price_provenance: Dict[str, Any] = Field(default_factory=dict)
    fx_provenance: Dict[str, Any] = Field(default_factory=dict)


class PortfolioHealthResponse(_StrictModel):
    as_of: date
    account_id: Optional[int] = None
    cost_method: Literal["fifo", "avg"]
    currency: str = Field(min_length=1, max_length=16)
    status: HealthStatus
    status_message: Optional[str] = Field(default=None, max_length=2000)
    score: Optional[ScoreFloat] = Field(
        default=None,
        description="Comparable deterministic score; null for incomplete coverage.",
    )
    partial_score: Optional[ScoreFloat] = Field(
        default=None,
        description=(
            "Fixed-denominator diagnostic estimate. Missing dimensions contribute zero; "
            "never compare this value with complete daily scores."
        ),
    )
    band: Optional[HealthBandName] = None
    coverage_ratio: UnitFloat
    comparable: bool
    disclaimer: str
    score_source: Literal["rules"] = "rules"
    llm_can_modify_score: Literal[False] = False
    formula_version: Literal["portfolio_health_v2"] = "portfolio_health_v2"
    weights: PortfolioHealthWeights
    effective_weights: PortfolioHealthEffectiveWeights
    bands: List[PortfolioHealthBand] = Field(default_factory=list, max_length=4)
    dimensions: PortfolioHealthDimensions
    unavailable_dimensions: List[
        Literal["concentration", "risk_exposure", "diversification", "pnl", "cash_ratio"]
    ] = Field(default_factory=list, max_length=5)
    insights: List[PortfolioHealthInsight] = Field(default_factory=list, max_length=64)
    data_quality: PortfolioHealthDataQuality
    inputs: PortfolioHealthInputs
    config: PortfolioHealthResolvedConfig
    provenance: PortfolioHealthProvenance
    persisted: bool = False
